src/irr/core/LRUcache.h

```diff
@@ -0,0 +1,131 @@
+#ifndef __LRU_CACHE_H_INCLUDED__
+#define __LRU_CACHE_H_INCLUDED__
+
+#include "irr/core/CFixedCapacityDoublyLinkedList.h"
+
+namespace irr {
+	namespace core {
+
+//Key-Value LRU cache
+//Stores fixed size amount of elements. 
+//When inserting new, prevents overflow by removing the least recently used element.
+template<typename Key, typename Value, typename MapHash = std::hash<Key>, typename MapEquals = std::equal_to<Key>>
+class LRUcache
+{
+	typedef std::pair<Key, Value> list_template_t;
+	_IRR_STATIC_INLINE_CONSTEXPR uint32_t invalid_iterator = Snode<list_template_t>::invalid_iterator;
+
+	DoublyLinkedList<list_template_t> m_list;
+	unordered_map<Key, uint32_t, MapHash, MapEquals> m_shortcut_map;
```

Comment:
why do you need the `Key` to live both in the linked list and the unordered map?

I know I gave a such example but it seems not necessary to duplicate it, @Crisspl, your opinion?

Reply:
mostly why does the key live in the linked list?

Reply:
P.S. Example I gave https://github.com/buildaworldnet/IrrlichtBAW/pull/487#discussion_r472068256

Reply:
It would be possible to only have one copy of `Key`, it would need a bit of a hack.

You'd need to change `unordered_map<Key, uint32_t, MapHash, MapEquals>` to `unordered_set<uint32_t, HashWrapper<MapHash>, EqualsWrapper<MapEquals>>`

Then you'd reserve the `invalid_iterator` to mean "input key to the function" (making a huge assertion here that `m_shortcut_map` can never contain `invalid_iterator`).

What would happen is that you could define your own `struct HashWrapper` and `struct EqualsWrapper` which thanks to C++11 could actually be made stateful and store the reference to the `m_list`, as well as `const Key* invalid_iterator_key;` and `std::size_t invalid_iterator_hash;`, you'd construct the wrappers in the constructor of `LRUcache` and pass them into the constructor of `m_shortcut_map`.

Then whenever you call `m_shortcut_map.find`, you'd first set the pointer to `invalid_iterator_key` and compute the `invalid_iterator_hash` for the stateful wrappers.

Then whenever the `operator()` of either the hash or equals gets called:
- If the value is `invalid_iterator` return the relevant state's hash or use the state's pointer for the `operator==`
- If the value is not `invalid_iterator` then fetch the relevant node from `m_list` use it's pair's `Key` member for hash or compare

Reply:
TODO: I don't like having a duplicate `Key` lingering around.

Also another solution would be to swap out the `Key` in unordered map for a `Key*` in the `unordered_map` (don't need a weird `unordered_set` anymore) however the hash and equals would still need to be wrapped to dereference the pointer.

The `Key*` pointer would need to point to the `SNode<>.value.first` so its basically almost a duplicate of the `uint32_t` field, at which point we degenerate into the `unordered_set` solution.

Reply:
I've changed the LRU cache to store `Key*` 😊👍